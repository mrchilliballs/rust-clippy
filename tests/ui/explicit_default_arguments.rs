#![warn(clippy::explicit_default_arguments)]

use std::marker::PhantomData;

// Test types
struct DatabaseError;
struct NetworkError;
struct ComplexStruct<A, B, C, D>(PhantomData<(A, B, C, D)>);

// Type aliases with defaults
type DbResult<T = ()> = Result<T, DatabaseError>;
type NetResult<T = &'static str> = Result<T, NetworkError>;
type Optional<T = i64> = Option<T>;
type ComplexThing<A, B, C = u32, D = f64> = ComplexStruct<A, B, C, D>;
type BoxedDefault<T = i128> = Box<T>;

// Module to test scoping
mod outer {
    pub type NestedResult<T = usize> = Result<T, ()>;
}

// Const declarations
const DB_CONST: DbResult<()> = Ok(());
//~^ explicit_default_arguments
const DB_OK: DbResult = Ok(());
const NET_CONST: NetResult<&str> = Ok("");
//~^ explicit_default_arguments
const NET_OK: NetResult = Ok("");

// Static declarations
static STATIC_DB: DbResult<()> = Ok(());
//~^ explicit_default_arguments
static STATIC_NET: NetResult = Ok("");
static OPTIONAL: Optional<i64> = Some(42);
//~^ explicit_default_arguments
static CUSTOM_OPT: Optional<f32> = Some(1.5);

// Associated types in traits
trait ExampleTrait {
    type AssocDb;
    type AssocNet;

    fn method() -> DbResult<()>;
    //~^ explicit_default_arguments
}

impl ExampleTrait for () {
    type AssocDb = DbResult<()>;
    //~^ explicit_default_arguments
    type AssocNet = NetResult;

    fn method() -> DbResult<()> {
        //~^ explicit_default_arguments
        Ok(())
    }
}

// Function signatures
fn db_function(arg: DbResult<()>) -> DbResult<()> {
    //~^ explicit_default_arguments
    //~| explicit_default_arguments
    arg
}

fn net_function(arg: NetResult) -> NetResult {
    arg
}

// Struct fields
struct User {
    db_field: DbResult<()>,
    //~^ explicit_default_arguments
    net_field: NetResult,
}

// Tuple struct
struct Response(
    DbResult<()>,
    //~^ explicit_default_arguments
    NetResult,
);

// Enum variants
enum ApiResponse {
    Success(DbResult<()>),
    //~^ explicit_default_arguments
    Failure(NetResult),
}

// Union fields
union DataHolder {
    db: std::mem::ManuallyDrop<DbResult<()>>,
    //~^ explicit_default_arguments
    net: std::mem::ManuallyDrop<NetResult>,
}

// Type aliases
type DbAlias = DbResult<()>;
//~^ explicit_default_arguments
type NetAlias = NetResult;

// Complex type scenarios
static COMPLEX_FULL: ComplexThing<i8, u8, u32, f64> = ComplexStruct(PhantomData);
//~^ explicit_default_arguments
static COMPLEX_PARTIAL: ComplexThing<i16, u16, u8> = ComplexStruct(PhantomData);

// Nested module type
static NESTED_RESULT: outer::NestedResult<usize> = Ok(42);
//~^ explicit_default_arguments

// Trait implementation with generics
impl<T> ExampleTrait for ComplexThing<T, T> {
    type AssocDb = DbResult<()>;
    //~^ explicit_default_arguments
    type AssocNet = NetResult;

    fn method() -> DbResult<()> {
        //~^ explicit_default_arguments
        Ok(())
    }
}

fn main() {
    // Local variables
    let a: DbResult<()> = Ok(());
    //~^ explicit_default_arguments
    let b: NetResult = Ok("");

    // Function pointers
    let f: fn(DbResult<()>) -> DbResult<()> = db_function;
    //~^ explicit_default_arguments
    //~| explicit_default_arguments

    // Expressions with std types
    let s = String::new();
    let v: Vec<String> = vec![s.clone()];
    let _o: Option<Vec<String>> = Some(v);

    // Box with default
    let boxed_int: BoxedDefault<i128> = Box::new(0);
    //~^ explicit_default_arguments
    let boxed_float: BoxedDefault<f64> = Box::new(0.0);
}
