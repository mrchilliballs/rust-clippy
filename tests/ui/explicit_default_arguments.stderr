error: use
  --> tests/ui/explicit_default_arguments.rs:23:17
   |
LL | const DB_CONST: DbResult<()> = Ok(());
   |                 ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`
   |
   = note: `-D clippy::explicit-default-arguments` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::explicit_default_arguments)]`

error: use
  --> tests/ui/explicit_default_arguments.rs:26:18
   |
LL | const NET_CONST: NetResult<&str> = Ok("");
   |                  ^^^^^^^^^^^^^^^ help: unnecessary generics, `&str` already is the default: `NetResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:31:19
   |
LL | static STATIC_DB: DbResult<()> = Ok(());
   |                   ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:34:18
   |
LL | static OPTIONAL: Optional<i64> = Some(42);
   |                  ^^^^^^^^^^^^^ help: unnecessary generics, `i64` already is the default: `Optional`

error: use
  --> tests/ui/explicit_default_arguments.rs:43:20
   |
LL |     fn method() -> DbResult<()>;
   |                    ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:48:20
   |
LL |     type AssocDb = DbResult<()>;
   |                    ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:52:20
   |
LL |     fn method() -> DbResult<()> {
   |                    ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:59:21
   |
LL | fn db_function(arg: DbResult<()>) -> DbResult<()> {
   |                     ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:59:38
   |
LL | fn db_function(arg: DbResult<()>) -> DbResult<()> {
   |                                      ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:71:15
   |
LL |     db_field: DbResult<()>,
   |               ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:78:5
   |
LL |     DbResult<()>,
   |     ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:85:13
   |
LL |     Success(DbResult<()>),
   |             ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:92:32
   |
LL |     db: std::mem::ManuallyDrop<DbResult<()>>,
   |                                ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:98:16
   |
LL | type DbAlias = DbResult<()>;
   |                ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:103:22
   |
LL | static COMPLEX_FULL: ComplexThing<i8, u8, u32, f64> = ComplexStruct(PhantomData);
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: unnecessary generics, [`u32`, `f64`] already are already the defaults: `ComplexThing<i8, u8>`

error: use
  --> tests/ui/explicit_default_arguments.rs:108:23
   |
LL | static NESTED_RESULT: outer::NestedResult<usize> = Ok(42);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: unnecessary generics, `usize` already is the default: `outer::NestedResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:113:20
   |
LL |     type AssocDb = DbResult<()>;
   |                    ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:117:20
   |
LL |     fn method() -> DbResult<()> {
   |                    ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:125:12
   |
LL |     let a: DbResult<()> = Ok(());
   |            ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:130:15
   |
LL |     let f: fn(DbResult<()>) -> DbResult<()> = db_function;
   |               ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:130:32
   |
LL |     let f: fn(DbResult<()>) -> DbResult<()> = db_function;
   |                                ^^^^^^^^^^^^ help: unnecessary generics, `()` already is the default: `DbResult`

error: use
  --> tests/ui/explicit_default_arguments.rs:140:20
   |
LL |     let boxed_int: BoxedDefault<i128> = Box::new(0);
   |                    ^^^^^^^^^^^^^^^^^^ help: unnecessary generics, `i128` already is the default: `BoxedDefault`

error: aborting due to 22 previous errors

